vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+using winrt::Windows::Storage::Streams::DataReader;
+using winrt::Windows::Storage::Streams::DataWriter;
+using winrt::Windows::Storage::Streams::UnicodeEncoding;
+
+namespace Microsoft::React
+{
+WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri)
+  : m_uri{ std::move(uri) }
+  , m_socket{}
+  , m_writer{ m_socket.OutputStream() }
+  , m_closeCode{ CloseCode::None }
+{
+  m_socket.MessageReceived({ this, &WinRTWebSocketResource::OnMessageReceived });
+  m_socket.Closed({ this, &WinRTWebSocketResource::OnClosed });
+
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::Untrusted);
+  m_socket.Control().IgnorableServerCertificateErrors()
+    .Append(winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult::InvalidName);
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
+}
+
+#pragma region Private members
+
+fire_and_forget WinRTWebSocketResource::PerformConnect()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    co_await m_socket.ConnectAsync(m_uri);
+
+    if (m_connectHandler)
+    {
+      m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    hresult_error e{ hr };
+    //OutputDebugString("WebSocket.connect failed (0x%8X) %ls\n", e);
+
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  // m_connectPerformed.Set();
+  m_connectPerformed.set_value();
+  m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    m_writer.WriteBytes(arr);
+
+    co_await m_writer.StoreAsync();
+
+    if (m_pingHandler)
+    {
+      m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(hresult_error(hr).message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  if (m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
```

Comment:
>wait [](start = 38, length = 4)

can we somehow do co_await here?

Reply:
Maybe. I haven't figured how yet.
The question is: WHAT would we be `co_await`ing for?

Reply:
I'm still relying on this `promise` for final synchronization, but I've also implemented coroutine-based sequencing.